### PR TITLE
S-94204 initial commit

### PR DESCRIPTION
**Pull request by rohithamroser**

### CODE REVIEW

File: gitlab/config.json [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+            "type": "input"
+        },
+        {
+            "title": "Description",
+            "name": "description",
+            "type": "input"
+        }
+    ],
+    "assets_properties": [
+
```
Comment (DhilipKumar0807) [inline gitlab/config.json:42]: remove this extra line

File: gitlab/config.json [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    "assets_properties": [
+        
+        {
+            "type":"custom",
+            "title":"organization",
+            "name":"organization",
+            "description":"To get the Group",
+            "api":"fetch_org"
+        },
+
```
Comment (DhilipKumar0807) [inline gitlab/config.json:50]: same here

File: gitlab/config.json [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+            "type": "project",
+            "title": " Projects",
+            "name": "projects",
+            "description": "To get the projects",
+            "is_render_treeview": true,
+            "max_projects": 10,
+            "is_fields_at_project_level": true,
+            "dependent_assets": ["organization"]
+        },
+
```
Comment (DhilipKumar0807) [inline gitlab/config.json:61]: same here and remove all the extra line after comma

File: gitlab/config.json [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+        },
+        
+        "missing_events": {
+            "enabled": true
+        },
+        "data_migration": {
+            "enabled": true
+        }
+    }
+}
```
Comment (DhilipKumar0807) [inline gitlab/config.json:91]: there should be a empty line at the end of the file

File: gitlab/default.py [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+        "title": "AS-Ticket-Create-Trigger",
+        "webhook_type": "ticket_created",
+        "conditions_value": "Create"
+    },
+    {
+        "title": "AS-Ticket-Update-Trigger",
+        "webhook_type": "ticket_updated",
+        "conditions_value": "Change"
+    }
+]
```
Comment (DhilipKumar0807) [inline gitlab/default.py:18]: Same here there should be a empty line at the end of the file

Comment (DhilipKumar0807) [inline gitlab/default.py:18]: Same here there should be a empty line at the end of the file

File: gitlab/fieldtype_expression.json [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+    },
+    "default": {
+        "payload_pattern": {
+            "create": "'$.object_attributes.%s' % (self.name)",
+            "update": "'$.changelog.items[?(@.field==\\'%s\\').toString]' % (self.name)"
+        },
+        "blank_value": null,
+        "reset_value": null
+    }
+}
```
Comment (DhilipKumar0807) [inline gitlab/fieldtype_expression.json:51]: Same here there should be a empty line at the end of the file

File: gitlab/mapping.py [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+            "TAGGER": "tagger",
+            "LOOKUP": "lookup",
+            "RELATION": "Relation"
+        }
+        fields = {
+            "title": {
+                "type": fields_type["TEXT"],
+                "system": "title"
+            }, 
+
```
Comment (DhilipKumar0807) [inline gitlab/mapping.py:78]: remove extra line spaces

File: gitlab/mapping.py [hunk 8]
```diff
@@ -0,0 +1,266 @@
...
+    def fetch_display_name(self):
+        return self.field_attr["title"]
+
+    def is_multivalue_field(self):
+        if self.field_attr["IsMultivalue"]:
+            return True
+        else:
+            return False
+    def _field_type_info(self, field_type, display_image, values=None, value_type=None):
```
Comment (DhilipKumar0807) [inline gitlab/mapping.py:46]: there should be a one-line gap for each function

File: gitlab/mapping.py [hunk 9]
```diff
@@ -0,0 +1,266 @@
...
+                "type": fields_type["INTEGER"],
+                "system": "weight"
+            },
+            "description": {
+                "type": fields_type["TEXT"],
+                "system": "description"
+            },
+            
+        }
+
```
Comment (DhilipKumar0807) [inline gitlab/mapping.py:94]: remove extra lines

File: gitlab/mapping.py [hunk 10]
```diff
@@ -0,0 +1,266 @@
...
+        
+        
+        attribute_type = fields[self.field_attr['type']]['type'].capitalize()
+
+        if attribute_type == 'Text':
+            return self._field_type_info(FieldTypes.TEXT,  FieldDisplayIcon.TEXT)
+        if attribute_type == 'Integer':
+            return self._field_type_info(FieldTypes.TEXT,  FieldDisplayIcon.TEXT)
+        elif attribute_type == 'Relation':
+            list =transformer_functions.get_field_value(self.instance_obj,details=self.instance_details,repo=self.fields_obj.project_info['display_name'],org = self.fields_obj.query_params["organization"]['display_name'] )
```
Comment (DhilipKumar0807) [inline gitlab/mapping.py:104]: equal to operator should have space around it

File: gitlab/mapping.py [hunk 11]
```diff
@@ -0,0 +1,266 @@
...
+    def connect(self):
+        return transformer_functions.connect(
+            self.instance_details
+        )
+    def fetch_sync_user(self):
+        user = self.instance_details["Username"]
+        return user
+    def fetch_projects(self):
+        org = self.query_params["organization"]["id"]
+        response_repos= transformer_functions.get_repos(self.instance_obj,org)
```
Comment (DhilipKumar0807) [inline gitlab/mapping.py:145]: after comma there should be space .get_repos(self.instance_obj,org)

File: gitlab/mapping.py [hunk 12]
```diff
@@ -0,0 +1,266 @@
...
+            
+                        "url" : webhook_url,
+                        "issues_events" : True
+
+                }  # Payload data to create single webhook
+       
+        for project in self.projects_info:
+            transformer_functions.webhooks(self.instance_obj,id = project['project']
+                                                      ,payload=payload)  # Creating webhook
+        #category_id = self.create_trigger_categories()  # Creating trigger category
```
Comment (DhilipKumar0807) [inline gitlab/mapping.py:189]: remove unused lines

File: gitlab/mapping.py [hunk 13]
```diff
@@ -0,0 +1,266 @@
...
+        for org in response_orgs:
+            orgs.append(
+                {
+                    'id': org['id'],
+                    "organization": org["name"],
+                    'display_name': org['name'],
+                }
+            )
+        return orgs
+    def connect(self):
```
Comment (DhilipKumar0807) [inline gitlab/mapping.py:136]: function should have a line space

File: gitlab/mapping.py [hunk 14]
```diff
@@ -0,0 +1,266 @@
...
+    def test_connection(self):
+        try:
+            return transformer_functions.check_connection(self.instance_obj,self.instance_details)
+        except Exception as ex:
+            raise as_exceptions.SanitizedPluginError("Unknown error connecting to GIthub Plugin integration system.", str(ex))
+
+class WebHook(BaseWebHook):
+    
+    def create_webhook(self, webhook_name, webhook_url,webhook_description,project_id):
+        hello = project_id
```
Comment (DhilipKumar0807) [inline gitlab/mapping.py:178]: remove this line

File: gitlab/transformer_functions.py [hunk 15]
```diff
@@ -0,0 +1,196 @@
...
+def ticfields(self):
+    return fields
+def connect(instance_details):
+   return ASyncRestApi(instance_details['url'],headers={
+                            "authorization": "Bearer {}".format(instance_details["token"]),
+                            "Accept": "application/json",
+                            "Content-Type": "application/json"
+                            })
+
+
```
Comment (DhilipKumar0807) [inline gitlab/transformer_functions.py:56]: Remove extra spaces and give a line space for each function

File: gitlab/transformer_functions.py [hunk 16]
```diff
@@ -0,0 +1,196 @@
+import base64
+
+from external_plugins.gitlab import default as DEFAULT 
+from agilitysync.external_lib.restapi import ASyncRestApi
+
+fields = [
+    {
+    "required": True,
+    "disabled field": False,
```
Comment (DhilipKumar0807) [inline gitlab/transformer_functions.py:9]: instead of "disabled field" change it disabled

File: gitlab/transformer_functions.py [hunk 17]
```diff
@@ -0,0 +1,196 @@
...
+    "customfield":False,
+    "raw_title": "title",
+    "title":"title",
```
Comment (DhilipKumar0807) [inline gitlab/transformer_functions.py:12]: there should space after ": " colon

File: gitlab/transformer_functions.py [hunk 18]
```diff
@@ -0,0 +1,196 @@
...
+                            "Accept": "application/json",
+                            "Content-Type": "application/json"
+                            })
+
+
+
+
+
+def check_connection(instance,instance_details):
+    instance_path = "repos"
```
Comment (DhilipKumar0807) [inline gitlab/transformer_functions.py:61]: constants have to be in full capital. Move this instance_path to top like - REPOS_ENDPOINT = "repos" and use the constants

File: gitlab/transformer_functions.py [hunk 19]
```diff
@@ -0,0 +1,196 @@
...
+            )
+
+    response = instance.get(path)
+
+    if instance_details["Username"] == response["username"]:
+        return "Connection to Gitlab server is successfull."
+    else:
+        return response["error"]
+def get_field_value(instance,details,repo,org):
+    instance_path = "milestones"
```
Comment (DhilipKumar0807) [inline gitlab/transformer_functions.py:76]: move to top as said in before comment

File: gitlab/sync.py [hunk 20]
```diff
@@ -0,0 +1,189 @@
...
+
+class Payload(BasePayload):
+
+    def fetch_project(self, event):
+        project = str(event["project"]["id"])
+        org = event['project']['namespace']
+        return org + "/" + project
+    
+    def fetch_asset(self, event):
+        return "Assettype-001"
```
Comment (DhilipKumar0807) [inline gitlab/sync.py:28]: what does "Assettype-001" represents?  the issue, repo? Can you make a  constant and get value from there GITLAB_ASSETS_ID =  {"issue": "Asset-001"}

File: gitlab/sync.py [hunk 21]
```diff
@@ -0,0 +1,189 @@
...
+
+
+class Event(BaseEvent):
+
+    def fetch_event_type(self):
+        event_type = self.event['object_attributes']['action']
+
+        if event_type in ('open'):
+            return EventTypes.CREATE
+        elif event_type == ('close'):
```
Comment (DhilipKumar0807) [inline gitlab/sync.py:41]: use anyone [in, ==] make it uniform If you are going to use the "in" operator use a tuple.

File: gitlab/sync.py [hunk 22]
```diff
@@ -0,0 +1,189 @@
...
+        category = []
+
+        event_type = self.event['object_attributes']["action"]
+
+        if event_type in ('open', 'close', 'update',):
+            category.append(EventCategory.WORKITEM)
+        
+        
+
+        ##if self.is_comment_updated(self.event["object_attributes"]["updated_at"], self.event["issue"]["comments_url"]):
```
Comment (DhilipKumar0807) [inline gitlab/sync.py:96]: remove the unused lines

File: gitlab/sync.py [hunk 23]
```diff
@@ -0,0 +1,189 @@
...
+
+        return create_fields
+
+    def create(self, sync_fields):
+        
+        
+        try: 
+            
+
+            ticket = transformer_functions.tickets(self.instance_object
```
Comment (DhilipKumar0807) [inline gitlab/sync.py:139]: check for space around the equal to operator

File: gitlab/config.json [hunk 24]
```diff
@@ -0,0 +1,91 @@
...
+    "instance_properties": [
+        {
+            "title": "Name",
+            "name": "name",
+            "description": "Name of the instance",
+            "type": "input"
+        },
+        {
+            "title": "Username",
+            "name": "Username",
```
Comment (DhilipKumar0807) [inline gitlab/config.json:16]: can you make this in lower case

File: gitlab/config.json [hunk 25]
```diff
@@ -0,0 +1,91 @@
...
+            "title": "Description",
+            "name": "description",
+            "type": "input"
+        }
+    ],
+    "assets_properties": [
+        
+        {
+            "type":"custom",
+            "title":"organization",
```
Comment (DhilipKumar0807) [inline gitlab/config.json:45]: Can you make titles first letter capital

File: gitlab/sync.py (hunk 20 above)
Comment (rohithamroser) [reply gitlab/sync.py:28]: issue

File: gitlab/config.json (hunk 24 above)
Comment (rohithamroser) [reply gitlab/config.json:16]: i used this on lot of places,so chnaginf it may rise a issue'

**Comment by rohithamroser** [comment]
changes are done

**Pull request closed without merging**